File: src/IECore/SimpleTypedData.cpp
//////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2007-2013, Image Engine Design Inc. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//
//     * Neither the name of Image Engine Design nor the names of any
//       other contributors to this software may be used to endorse or
//       promote products derived from this software without specific prior
//       written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//////////////////////////////////////////////////////////////////////////

#include "IECore/SimpleTypedData.h"
#include "IECore/Export.h"
#include "IECore/TypedData.inl"

#include <iostream>

namespace IECore
{

static IndexedIO::EntryID g_valueEntry("value");

LongDataAlias::TypeDescription<IntData> LongDataAlias::m_typeDescription( LongDataTypeId, "LongData" );

#define IE_CORE_DEFINEBASETYPEDDATAIOSPECIALISATION( TNAME, N, FALLBACKNAME )	\
	template<> \
	void TNAME::save( SaveContext *context ) const \
	{ \
		Data::save( context ); \
		assert( baseSize() == N ); \
		IndexedIO *container = context->rawContainer(); \
		container->write( g_valueEntry, TNAME::baseReadable(), TNAME::baseSize() ); \
	} \
	\
	template<> \
	void TNAME::load( LoadContextPtr context ) \
	{ \
		Data::load( context ); \
		assert( ( sizeof( TNAME::ValueType ) / sizeof( TNAME::BaseType ) ) == N ); \
		TNAME::BaseType *p = TNAME::baseWritable(); \
		try \
		{ \
			const IndexedIO *container = context->rawContainer(); \
			container->read( g_valueEntry, p, N ); \
		} \
		catch( ... ) \
		{ \
			unsigned int v = 0;	\
			ConstIndexedIOPtr container = context->container( FALLBACKNAME::staticTypeName(), v ); \
			container->read( g_valueEntry, p, N ); \
		} \
	}

#define IECORE_DEFINE_ZERO_INITIALISED_CONSTRUCTOR( TNAME ) \
	template<> \
	TNAME::TypedData() \
		:	m_data( ValueType( 0 ) ) \
	{ \
	}

#define IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( TNAME, TID, N ) \
	IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( TNAME, TID ) \
	IE_CORE_DEFINEBASETYPEDDATAIOSPECIALISATION( TNAME, N, TNAME ) \

#define IE_CORE_DEFINEIMATHGEOMETRICTYPEDDATASPECIALISATION( TNAME, TID, BTID, N ) \
	IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( TNAME ## Base, BTID ) \
	IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( TNAME, TID ) \
	IE_CORE_DEFINEBASETYPEDDATAIOSPECIALISATION( TNAME ## Base, N, TNAME ) \

IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( BoolData, BoolDataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( FloatData, FloatDataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( DoubleData, DoubleDataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( IntData, IntDataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( UIntData, UIntDataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( CharData, CharDataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( UCharData, UCharDataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( HalfData, HalfDataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( ShortData, ShortDataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( UShortData, UShortDataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( Int64Data, Int64DataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( UInt64Data, UInt64DataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( StringData, StringDataTypeId )
IECORE_RUNTIMETYPED_DEFINETEMPLATESPECIALISATION( InternedStringData, InternedStringDataTypeId )

IE_CORE_DEFINEIMATHGEOMETRICTYPEDDATASPECIALISATION( V2iData, V2iDataTypeId, V2iDataBaseTypeId, 2 )
IE_CORE_DEFINEIMATHGEOMETRICTYPEDDATASPECIALISATION( V3iData, V3iDataTypeId, V3iDataBaseTypeId, 3 )
IE_CORE_DEFINEIMATHGEOMETRICTYPEDDATASPECIALISATION( V2fData, V2fDataTypeId, V2fDataBaseTypeId, 2 )
IE_CORE_DEFINEIMATHGEOMETRICTYPEDDATASPECIALISATION( V3fData, V3fDataTypeId, V3fDataBaseTypeId, 3 )
IE_CORE_DEFINEIMATHGEOMETRICTYPEDDATASPECIALISATION( V2dData, V2dDataTypeId, V2dDataBaseTypeId, 2 )
IE_CORE_DEFINEIMATHGEOMETRICTYPEDDATASPECIALISATION( V3dData, V3dDataTypeId, V3dDataBaseTypeId, 3 )

IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( Color3fData, Color3fDataTypeId, 3 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( Color4fData, Color4fDataTypeId, 4 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( Color3dData, Color3dDataTypeId, 3 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( Color4dData, Color4dDataTypeId, 4 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( Box2iData, Box2iDataTypeId, 4 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( Box3iData, Box3iDataTypeId, 6 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( Box2fData, Box2fDataTypeId, 4 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( Box3fData, Box3fDataTypeId, 6 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( Box2dData, Box2dDataTypeId, 4 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( Box3dData, Box3dDataTypeId, 6 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( M33fData, M33fDataTypeId, 9 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( M33dData, M33dDataTypeId, 9 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( M44fData, M44fDataTypeId, 16 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( M44dData, M44dDataTypeId, 16 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( QuatfData, QuatfDataTypeId, 4 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( QuatdData, QuatdDataTypeId, 4 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( LineSegment3fData, LineSegment3fDataTypeId, 6 )
IE_CORE_DEFINEIMATHTYPEDDATASPECIALISATION( LineSegment3dData, LineSegment3dDataTypeId, 6 )

IECORE_DEFINE_ZERO_INITIALISED_CONSTRUCTOR( V2iDataBase )
IECORE_DEFINE_ZERO_INITIALISED_CONSTRUCTOR( V3iDataBase )
IECORE_DEFINE_ZERO_INITIALISED_CONSTRUCTOR( V2fDataBase )
IECORE_DEFINE_ZERO_INITIALISED_CONSTRUCTOR( V3fDataBase )
IECORE_DEFINE_ZERO_INITIALISED_CONSTRUCTOR( V2dDataBase )
IECORE_DEFINE_ZERO_INITIALISED_CONSTRUCTOR( V3dDataBase )
IECORE_DEFINE_ZERO_INITIALISED_CONSTRUCTOR( Color3fData )
IECORE_DEFINE_ZERO_INITIALISED_CONSTRUCTOR( Color4fData )
IECORE_DEFINE_ZERO_INITIALISED_CONSTRUCTOR( Color3dData )
IECORE_DEFINE_ZERO_INITIALISED_CONSTRUCTOR( Color4dData )

template<>
void StringData::memoryUsage( Object::MemoryAccumulator &accumulator ) const
{
	Data::memoryUsage( accumulator );
	accumulator.accumulate( readable().capacity() );
}

template<>
void TypedData<InternedString>::save( SaveContext *context ) const
{
	Data::save( context );
	IndexedIO *container = context->rawContainer();
	container->write( g_valueEntry, readable().value() );
}

template<>
void TypedData<InternedString>::load( LoadContextPtr context )
{
	Data::load( context );
	std::string v;
	const IndexedIO *container = context->rawContainer();
	container->read( g_valueEntry, v );
	writable() = v;
}

template<>
void TypedData<bool>::save( SaveContext *context ) const
{
	Data::save( context );
	IndexedIO *container = context->rawContainer();
	unsigned char c = readable();
	container->write( g_valueEntry, c );
}

template<>
void TypedData<bool>::load( LoadContextPtr context )
{
	Data::load( context );
	unsigned char c;
	try
	{
		// optimised format for new files
		const IndexedIO *container = context->rawContainer();
		container->read( g_valueEntry, c );
	}
	catch( ... )
	{
		// backwards compatibility with old files
		unsigned int v = 0;
		ConstIndexedIOPtr container = context->container( staticTypeName(), v );
		container->read( g_valueEntry, c );
	}

	writable() = c;
}

template<>
void TypedData<short>::save( SaveContext *context ) const
{
	Data::save( context );
	IndexedIO *container = context->rawContainer();
	int c = readable();
	container->write( g_valueEntry, c );
}

template<>
void TypedData<short>::load( LoadContextPtr context )
{
	Data::load( context );
	int c;
	try
	{
		// optimised format for new files
		const IndexedIO *container = context->rawContainer();
		container->read( g_valueEntry, c );
	}
	catch( ... )
	{
		// backwards compatibility with old files
		unsigned int v = 0;
		ConstIndexedIOPtr container = context->container( staticTypeName(), v );
		container->read( g_valueEntry, c );
	}

	writable() = static_cast<short>( c );
}

template<>
void TypedData<unsigned short>::save( SaveContext *context ) const
{
	Data::save( context );
	IndexedIO *container = context->rawContainer();
	unsigned int c = readable();
	container->write( g_valueEntry, c );
}

template<>
void TypedData<unsigned short>::load( LoadContextPtr context )
{
	Data::load( context );
	unsigned int c;
	try
	{
		// optimised format for new files
		const IndexedIO *container = context->rawContainer();
		container->read( g_valueEntry, c );
	}
	catch( ... )
	{
		// backwards compatibility with old files
		unsigned int v = 0;
		ConstIndexedIOPtr container = context->container( staticTypeName(), v );
		container->read( g_valueEntry, c );
	}

	writable() = static_cast<unsigned short>( c );
}

template<>
TypedData<LineSegment3f>::TypedData()
	:	m_data( LineSegment3f( Imath::V3f( 0 ), Imath::V3f( 1, 0, 0 ) ) )
{
}

template<>
TypedData<LineSegment3d>::TypedData()
	:	m_data( LineSegment3d( Imath::V3d( 0 ), Imath::V3d( 1, 0, 0 ) ) )
{
}

template<>
void SimpleDataHolder<LineSegment3f>::hash( MurmurHash &h ) const
{
	h.append( readable().p0 );
	h.append( readable().p1 );
}

template<>
void SimpleDataHolder<LineSegment3d>::hash( MurmurHash &h ) const
{
	h.append( readable().p0 );
	h.append( readable().p1 );
}

template class IECORE_API TypedData<bool>;
template class IECORE_API TypedData<float>;
template class IECORE_API TypedData<double>;
template class IECORE_API TypedData<int>;
template class IECORE_API TypedData<unsigned int>;
template class IECORE_API TypedData<char>;
template class IECORE_API TypedData<unsigned char>;
template class IECORE_API TypedData<short>;
template class IECORE_API TypedData<unsigned short>;
template class IECORE_API TypedData<int64_t>;
template class IECORE_API TypedData<uint64_t>;
template class IECORE_API TypedData<std::string>;
template class IECORE_API TypedData<InternedString>;
template class IECORE_API TypedData<half>;

template class IECORE_API TypedData<Imath::V2i>;
template class IECORE_API TypedData<Imath::V3i>;
template class IECORE_API TypedData<Imath::V2f>;
template class IECORE_API TypedData<Imath::V3f>;
template class IECORE_API TypedData<Imath::V2d>;
template class IECORE_API TypedData<Imath::V3d>;

template class IECORE_API GeometricTypedData<Imath::V2i>;
template class IECORE_API GeometricTypedData<Imath::V3i>;
template class IECORE_API GeometricTypedData<Imath::V2f>;
template class IECORE_API GeometricTypedData<Imath::V3f>;
template class IECORE_API GeometricTypedData<Imath::V2d>;
template class IECORE_API GeometricTypedData<Imath::V3d>;

template class IECORE_API TypedData<Imath::Color3f>;
template class IECORE_API TypedData<Imath::Color4f>;
template class IECORE_API TypedData<Imath::Color3<double> >;
template class IECORE_API TypedData<Imath::Color4<double> >;
template class IECORE_API TypedData<Imath::Box2i>;
template class IECORE_API TypedData<Imath::Box3i>;
template class IECORE_API TypedData<Imath::Box2f>;
template class IECORE_API TypedData<Imath::Box3f>;
template class IECORE_API TypedData<Imath::Box2d>;
template class IECORE_API TypedData<Imath::Box3d>;
template class IECORE_API TypedData<Imath::M33f>;
template class IECORE_API TypedData<Imath::M33d>;
template class IECORE_API TypedData<Imath::M44f>;
template class IECORE_API TypedData<Imath::M44d>;
template class IECORE_API TypedData<Imath::Quatf>;
template class IECORE_API TypedData<Imath::Quatd>;
template class IECORE_API TypedData<LineSegment3f>;
template class IECORE_API TypedData<LineSegment3d>;

} // namespace IECore

File: test/IECore/SimpleTypedData.py
##########################################################################
#
#  Copyright (c) 2007-2013, Image Engine Design Inc. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#
#     * Neither the name of Image Engine Design nor the names of any
#       other contributors to this software may be used to endorse or
#       promote products derived from this software without specific prior
#       written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################

"""Unit test for SimpleTypedData binding"""

import os
import math
import unittest
import random
import six
import imath

import IECore

class SimpleTypedDataTest(unittest.TestCase):

	def __toLong( self, o ) :

		if six.PY3 :
			return int( o )
		else :
			return long( o )

	def testIntData(self):
		"""Test IntData"""
		a = IECore.IntData()
		self.assert_(type(a) is IECore.IntData)
		b = IECore.IntData(1)
		c = b.copy()
		self.assertEqual(b.value, 1)
		self.assertEqual(int(b), 1)
		b.value = 2
		self.assertEqual(b.value, 2)
		a = IECore.IntData(2)
		self.assert_(c < b)
		self.assert_(not c > b)
		self.assert_(a == b)
		self.assert_(not a == c)
		self.assertEqual( str( IECore.IntData() ), "0" )
		self.assertEqual( repr( IECore.IntData() ), "IECore.IntData( 0 )" )
		self.failUnless( IECore.IntData.hasBase() )

	def testUIntData(self):
		"""Test UIntData"""
		a = IECore.UIntData()
		self.assert_(type(a) is IECore.UIntData)
		b = IECore.UIntData(1)
		c = b.copy()
		self.assertEqual(b.value, 1)
		self.assertEqual( self.__toLong( b ), 1 )
		b.value = 2
		self.assertEqual(b.value, 2)
		a = IECore.UIntData(2)
		self.assert_(c < b)
		self.assert_(not c > b)
		self.assert_(a == b)
		self.assert_(not a == c)
		self.assertEqual( str( IECore.UIntData() ), "0" )
		self.assertEqual( repr( IECore.UIntData() ), "IECore.UIntData( 0 )" )
		self.failUnless( IECore.UIntData.hasBase() )

	def testFloatData(self):
		"""Test FloatData"""
		a = IECore.FloatData()
		self.assert_(type(a) is IECore.FloatData)
		b = IECore.FloatData(1)
		c = b.copy()
		self.assertEqual(b.value, 1)
		self.assertEqual(float(b), 1)
		b.value = 2
		self.assertEqual(b.value, 2)
		a = IECore.FloatData(2)
		self.assert_(c < b)
		self.assert_(not c > b)
		self.assert_(a == b)
		self.assert_(not a == c)
		self.assertEqual( str( IECore.FloatData() ), "0" )
		self.assertEqual( repr( IECore.FloatData() ), "IECore.FloatData( 0 )" )
		self.failUnless( IECore.FloatData.hasBase() )

	def testDoubleData(self):
		"""Test DoubleData"""
		a = IECore.DoubleData()
		self.assert_(type(a) is IECore.DoubleData)
		b = IECore.DoubleData(1)
		c = b.copy()
		self.assertEqual(b.value, 1)
		self.assertEqual(float(b), 1)
		b.value = 2
		self.assertEqual(b.value, 2)
		a = IECore.DoubleData(2)
		self.assert_(c < b)
		self.assert_(not c > b)
		self.assert_(a == b)
		self.assert_(not a == c)
		self.assertEqual( str( IECore.DoubleData() ), "0" )
		self.assertEqual( repr( IECore.DoubleData() ), "IECore.DoubleData( 0 )" )
		self.failUnless( IECore.DoubleData.hasBase() )

	def testCharData(self):
		"""Test CharData"""
		a = IECore.CharData()
		self.assert_(type(a) is IECore.CharData)
		b = IECore.CharData('1')
		c = b.copy()
		self.assertEqual(b.value, '1')
		b.value = '2'
		self.assertEqual(b.value, '2')
		a = IECore.CharData('2')
		self.assert_(c < b)
		self.assert_(not c > b)
		self.assert_(a == b)
		self.assert_(not a == c)
		self.assertEqual( str( IECore.CharData() ), "0" )
		self.assertEqual( repr( IECore.CharData() ), "IECore.CharData( 0 )" )
		self.failUnless( IECore.CharData.hasBase() )

	def testUCharData(self):
		"""Test UCharData"""
		a = IECore.UCharData()
		self.assert_(type(a) is IECore.UCharData)
		b = IECore.UCharData(1)
		c = b.copy()
		self.assertEqual(b.value, 1)
		self.assertEqual(int(b), 1)
		b.value = 2
		self.assertEqual(b.value, 2)
		a = IECore.UCharData(2)
		self.assert_(c < b)
		self.assert_(not c > b)
		self.assert_(a == b)
		self.assert_(not a == c)
		self.assertEqual( str( IECore.UCharData() ), "0" )
		self.assertEqual( repr( IECore.UCharData() ), "IECore.UCharData( 0 )" )
		self.failUnless( IECore.UCharData.hasBase() )

	def testHalfData(self):
		"""Test HalfData"""
		a = IECore.HalfData()
		self.assert_(type(a) is IECore.HalfData)
		b = IECore.HalfData(1)
		c = b.copy()
		self.assertEqual(b.value, 1)
		self.assertEqual(float(b), 1)
		b.value = 2
		self.assertEqual(b.value, 2)
		a = IECore.HalfData(2)
		self.assert_(c < b)
		self.assert_(not c > b)
		self.assert_(a == b)
		self.assert_(not a == c)
		self.assertEqual( str( IECore.HalfData() ), "0" )
		self.assertEqual( repr( IECore.HalfData() ), "IECore.HalfData( 0 )" )
		self.failUnless( IECore.HalfData.hasBase() )

	def testShortData(self):
		"""Test ShortData"""
		a = IECore.ShortData()
		self.assert_(type(a) is IECore.ShortData)
		b = IECore.ShortData(1)
		c = b.copy()
		self.assertEqual(b.value, 1)
		self.assertEqual(int(b), 1)
		b.value = 2
		self.assertEqual(b.value, 2)
		a = IECore.ShortData(2)
		self.assert_(c < b)
		self.assert_(not c > b)
		self.assert_(a == b)
		self.assert_(not a == c)
		self.assertEqual( str( IECore.ShortData() ), "0" )
		self.assertEqual( repr( IECore.ShortData() ), "IECore.ShortData( 0 )" )
		self.failUnless( IECore.ShortData.hasBase() )

	def testUShortData(self):
		"""Test UShortData"""
		a = IECore.UShortData()
		self.assert_(type(a) is IECore.UShortData)
		b = IECore.UShortData(1)
		c = b.copy()
		self.assertEqual(b.value, 1)
		self.assertEqual(int(b), 1)
		b.value = 2
		self.assertEqual(b.value, 2)
		a = IECore.UShortData(2)
		self.assert_(c < b)
		self.assert_(not c > b)
		self.assert_(a == b)
		self.assert_(not a == c)
		self.assertEqual( str( IECore.UShortData() ), "0" )
		self.assertEqual( repr( IECore.UShortData() ), "IECore.UShortData( 0 )" )
		self.failUnless( IECore.UShortData.hasBase() )

	def testInt64Data(self):
		"""Test Int64Data"""
		a = IECore.Int64Data()
		self.assert_(type(a) is IECore.Int64Data)
		b = IECore.Int64Data(1)
		c = b.copy()
		self.assertEqual(b.value, 1)
		self.assertEqual( self.__toLong( b ), 1 )
		b.value = 2
		self.assertEqual(b.value, 2)
		a = IECore.Int64Data(2)
		self.assert_(c < b)
		self.assert_(not c > b)
		self.assert_(a == b)
		self.assert_(not a == c)
		self.assertEqual( str( IECore.Int64Data() ), "0" )
		self.assertEqual( repr( IECore.Int64Data() ), "IECore.Int64Data( 0 )" )
		self.failUnless( IECore.Int64Data.hasBase() )

	def testUInt64Data(self):
		"""Test UInt64Data"""
		a = IECore.UInt64Data()
		self.assert_(type(a) is IECore.UInt64Data)
		b = IECore.UInt64Data(1)
		c = b.copy()
		self.assertEqual(b.value, 1)
		self.assertEqual( self.__toLong( b ), 1 )
		b.value = 2
		self.assertEqual(b.value, 2)
		a = IECore.UInt64Data(2)
		self.assert_(c < b)
		self.assert_(not c > b)
		self.assert_(a == b)
		self.assert_(not a == c)
		self.assertEqual( str( IECore.UInt64Data() ), "0" )
		self.assertEqual( repr( IECore.UInt64Data() ), "IECore.UInt64Data( 0 )" )
		self.failUnless( IECore.UInt64Data.hasBase() )

	def testImathVecTypes(self):

		types = [
			[IECore.V2fData, imath.V2f],
			[IECore.V3fData, imath.V3f],
			[IECore.V2dData, imath.V2d],
			[IECore.V3dData, imath.V3d],
			[IECore.Color3fData, imath.Color3f],
			[IECore.Color4fData, imath.Color4f],
		]

		for t, vt in types :

			self.assertEqual( t(), t( vt( 0 ) ) )

			v = vt( 10 )
			self.assertEqual( " ".join( ["10"] * vt.dimensions() ), str( t( v ) ) )
			self.assertEqual( "IECore." + t.__name__ + "( " + "imath." + vt.__name__ + "( " + ", ".join( ["10"] * vt.dimensions() ) + " ) )", repr( t( v ) ) )
			self.failUnless( t.hasBase() )

			for i in range( 0, 1000 ) :
				v = vt( random.uniform( -100, 100 ) )
				d = t( v )
				self.assertEqual( d.value, v )
				v2 = vt( random.uniform( -100, 100 ) )
				d.value = v2
				self.assertEqual( d.value, v2 )

	def testImathBoxTypes(self):

		types = [
			[IECore.Box2fData, imath.Box2f, imath.V2f],
			[IECore.Box3fData, imath.Box3f, imath.V3f],
			[IECore.Box2dData, imath.Box2d, imath.V2d],
			[IECore.Box3dData, imath.Box3d, imath.V3d] ]

		for t, bt, vt in types :

			v = vt( 1 )
			b = bt( v, v )
			self.assertEqual( " ".join( ["1"]*vt.dimensions()*2 ), str( t( b ) ) )
			vr = "imath." + vt.__name__ + "( " + ", ".join( ["1"]*vt.dimensions() ) + " )"
			br = "imath." + bt.__name__ + "( " + vr + ", " + vr + " )"
			self.assertEqual( "IECore." + t.__name__ + "( " + br + " )", repr( t( b ) ) )
			self.failUnless( t.hasBase() )

			for i in range( 0, 1000 ) :

				b = bt()
				for j in range( 0, 10 ) :

					b.extendBy( vt( random.uniform( -100, 100 ) ) )

				v = t( b )
				self.assertEqual( v.value, b )
				b2 = bt()
				v.value = b2
				self.assertEqual( v.value, b2 )

	def testInternedStringData( self ) :

		s = IECore.InternedStringData( "i" )
		self.assertEqual( str( s ), "i" )
		self.assertEqual( repr(s ), 'IECore.InternedStringData( "i" )' )

		m = IECore.MemoryIndexedIO( IECore.CharVectorData(), [], IECore.IndexedIO.OpenMode.Append )
		s.save( m, "o" )

		s2 = IECore.Object.load( m, "o" )

		self.assertEqual( s2.value.value(), "i" )
		self.assertEqual( s, s2 )

	def testStringDataRepr( self ) :

		for s in [
			"a\nb",
			"a'",
			'"',
		] :
			d = IECore.StringData( s )
			self.assertEqual( eval( repr( d ) ), d )

	def testComparison( self ) :

		for i in range( 0, 1000 ) :

			self.assertTrue( IECore.IntData( 10 ) < IECore.IntData( 20 ) )
			self.assertFalse( IECore.IntData( 20 ) < IECore.IntData( 10 ) )

			self.assertTrue( IECore.IntData( 11 ) > IECore.IntData( 10 ) )
			self.assertFalse( IECore.IntData( 10 ) > IECore.IntData( 10 ) )

			self.assertTrue( IECore.IntData( 11 ) >= IECore.IntData( 11 ) )
			self.assertTrue( IECore.IntData( 12 ) >= IECore.IntData( 11 ) )
			self.assertFalse( IECore.IntData( 9 ) >= IECore.IntData( 10 ) )

			self.assertTrue( IECore.IntData( 11 ) <= IECore.IntData( 11 ) )
			self.assertTrue( IECore.IntData( 10 ) <= IECore.IntData( 11 ) )
			self.assertFalse( IECore.IntData( 11 ) <= IECore.IntData( 10 ) )

	def testEqualityDoesntThrow( self ) :

		self.assertFalse( IECore.IntData( 100 ) == None )
		self.assertTrue( IECore.IntData( 100 ) != None )

class BoolDataTest( unittest.TestCase ) :

	def test( self ) :

		a = IECore.BoolData()
		self.assert_(type(a) is IECore.BoolData)
		b = IECore.IntData(True)
		c = b.copy()
		self.assertEqual(b.value, True)
		b.value = False
		self.assertEqual(b.value, False)
		self.assertEqual(c.value, True)

		self.failUnless( IECore.BoolData.hasBase() )

	def testStreaming( self ) :

		o = IECore.BoolData( True )
		self.assertEqual( o.value, True )

		iface = IECore.IndexedIO.create( "test/IECore/o.fio", IECore.IndexedIO.OpenMode.Write )

		o.save( iface, "test" )
		oo = IECore.Object.load( iface, "test" )
		self.assertEqual( o, oo )

		o = IECore.BoolData( False )
		self.assertEqual( o.value, False )

		o.save( iface, "test" )
		oo = IECore.Object.load( iface, "test" )
		self.assertEqual( o, oo )

	def testLineSegmentData( self ) :

		for vt, dt in [
			( imath.V3f, IECore.LineSegment3fData ),
			( imath.V3d, IECore.LineSegment3dData ),
		] :

			d = dt()
			self.assertEqual( d.value.p0, vt( 0, 0, 0 ) )
			self.assertEqual( d.value.p1, vt( 1, 0, 0 ) )

	def tearDown( self ) :

		if os.path.isfile("test/IECore/o.fio"):
			os.remove("test/IECore/o.fio")

if __name__ == "__main__":
    unittest.main()

